subsys/audio_module/audio_module.c

```diff
@@ -0,0 +1,1065 @@
+
+		return 0;
+	}
+
+	k_mutex_lock(&handle->dest_mutex, K_FOREVER);
+
+	k_sem_init(&handle->sem, handle->dest_count, handle->dest_count);
+
+	SYS_SLIST_FOR_EACH_CONTAINER(&handle->hdl_dest_list, handle_to, node) {
+		ret = data_tx(handle, handle_to, block, &block_release_cb);
+		if (ret) {
+			LOG_DBG("Failed to send to module %s from %s, ret %d", handle_to->name,
+				handle->name, ret);
+			k_sem_take(&handle->sem, K_NO_WAIT);
+		}
+	}
+
+	if (handle->data_tx && handle->thread.msg_tx) {
+		ret = send_to_tx_fifo(handle, block);
+		if (ret) {
+			LOG_DBG("Failed to send block on module %s TX message queue", handle->name);
+			k_sem_take(&handle->sem, K_NO_WAIT);
+		}
+	}
+
+	k_mutex_unlock(&handle->dest_mutex);
+
+	return 0;
+}
+
+/**
+ * @brief The thread that receives data from outside the system and passes it into the audio system.
+ *
+ * @param handle  The handle for this modules instance
+ *
+ * @return 0 if successful, error value
+ */
+static int module_thread_input(struct audio_module_handle *handle, void *p2, void *p3)
+{
+	int ret;
+	struct audio_data block;
+	void *data;
+
+	if (handle == NULL) {
+		LOG_DBG("Module task has NULL handle");
+		return -EINVAL;
+	}
+
+	/* Execute thread */
+	while (1) {
+		data = NULL;
+
+		if (handle->description->functions->data_process != NULL) {
+			/* Get a new output buffer */
+			ret = k_mem_slab_alloc(handle->thread.data_slab, (void **)&data, K_FOREVER);
+			if (ret) {
+				LOG_DBG("No free data for module %s, ret %d", handle->name, ret);
+				continue;
+			}
+
+			/* Configure new audio block */
+			block.data = data;
+			block.data_size = handle->thread.data_size;
+
+			/* Process the input audio block */
+			ret = handle->description->functions->data_process(
+				(struct audio_module_handle_private *)handle, NULL, &block);
+			if (ret) {
+				k_mem_slab_free(handle->thread.data_slab, (void **)(&data));
+
+				LOG_DBG("Data process error in module %s, ret %d", handle->name,
+					ret);
+				continue;
+			}
+
+			/* Send input audio block to next module(s) */
+			send_to_connected(handle, &block);
+		} else {
+			LOG_DBG("No process function for module %s, no input", handle->name);
+		}
+	}
+
+	return 0;
+}
+
+/**
+ * @brief The thread that processes inputs and outputs them out of the audio system.
+ *
+ * @param handle  The handle for this modules instance
+ *
+ * @return 0 if successful, error value
+ */
+static int module_thread_output(struct audio_module_handle *handle, void *p2, void *p3)
+{
+	int ret;
+
+	struct audio_module_message *msg_rx;
+	size_t size;
+
+	if (handle == NULL) {
+		LOG_DBG("Module task has NULL handle");
+		return -EINVAL;
+	}
+
+	/* Execute thread */
+	while (1) {
+		msg_rx = NULL;
+
+		ret = data_fifo_pointer_last_filled_get(handle->thread.msg_rx, (void **)&msg_rx,
+							&size, K_FOREVER);
+		if (ret) {
+			LOG_DBG("No new data for module %s, ret %d", handle->name, ret);
+			continue;
+		}
+
+		if (handle->description->functions->data_process != NULL) {
```

Comment:
To reduce the amount of indentations, I recommend checking 
```suggestion
		if (handle->description->functions->data_process == NULL) { 
		 LOG_ERR("...");
		}
```
then, the main part of the code will be indented one level less. This goes for several places.

Reply:
I have moved this outside of the while loop as no need to check every iteration.